notebooks/nearest_neighbors.livemd

```diff
@@ -0,0 +1,255 @@
+
+```elixir
+knn = knn[[.., 1..-1//1]]
+age_groups = Nx.take(y, knn)
+
+seniors =
+  Nx.divide(
+    Nx.sum(Nx.multiply(Nx.equal(age_groups, 1), Nx.new_axis(Nx.equal(y, 1), 1))),
+    Nx.sum(Nx.equal(y, 1))
+  )
+  |> Nx.divide(num_neighbors - 1)
+
+adults =
+  Nx.divide(
+    Nx.sum(Nx.multiply(Nx.equal(age_groups, 0), Nx.new_axis(Nx.equal(y, 0), 1))),
+    Nx.sum(Nx.equal(y, 0))
+  )
+  |> Nx.divide(num_neighbors - 1)
+
+{seniors, adults}
+```
+
+Our intuition was correct: for seniors, 28% of their nearest neighbors in terms of medical results are also seniors, while for adults, nearly 86% of their nearest neighbors are also adults. This indicates that age should always be considered to ensure proper treatment when assessing medical results and drawing conclusions.
+
+## How to compute Approximated Nearest Neighbors - LargeVis
+
+In the previous example, KDTree was a perfect choice for computing K-Nearest Neighbors. However, this approach doesn't always work effectively. When the number of features grows, the efficiency of KDTree decreases significantly, so it's best suited for data with a smaller number of features (<15), like in the previous example. For data with many samples but also many features, a different approach is required. Here, algorithms that return an approximated version of nearest neighbors are needed. These approximated algorithms may not give exact solutions, but the deviation is usually minor and accepted as a tradeoff for faster computation. An example of an algorithm that computes approximated nearest neighbors is LargeVis.
+
+LargeVis is an approximated nearest neighbors algorithm that uses the concept of a Random Projection Forest. A Random Projection Forest consists of multiple trees. Each tree is built using a divide-and-conquer approach. The entire dataset is used to start the construction, and at each node, the data is projected onto a random hyperplane and split as follows: points with a projection smaller than or equal to the median go into the left subtree, while those with a projection greater than the median go into the right subtree. This process is then recursively applied to the left and right subtrees.
+
+This structure allows us to quickly create an approximated nearest neighbors graph, which can then be used to find the nearest neighbors efficiently. To better understand this, below is a figure illustrating the process of creating a tree in a random projection forest.
+
+| ![Random Projection Forest](http://www.math.umassd.edu/~dyan/rpKNN.png?attredirects=0) |
+| :------------------------------------------------------------------------------------: |
+
+Figure 3: A process of creating a tree in Random Projection Forest. Source: http://www.math.umassd.edu/~dyan/rpKNN.png?attredirects=0
```

Comment:
Not sure about this figure. The reason is that random projection trees in `Scholar` always split the data in half. Perhaps the one from https://cseweb.ucsd.edu/~dasgupta/papers/rptree-stoc.pdf or https://cseweb.ucsd.edu/~dasgupta/papers/exactnn-algo.pdf might be a better idea?

Reply:
Are there any photos of these on the internet that I could link to?

Reply:
I will include the image from the first paper that shows how splitting looks like in the case of kdtree and rpf, but I will leave the original image and mention that actual splitting in Scholar is implemented differently.

Reply:
That might be a better idea.

Reply:
I don't think it's that important, so feel free to leave the existing figure if you want. :)